File: examples/python/table/agg/handlers/end_hand.py
"""Handler for EndHand command."""

import sys
from pathlib import Path
from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

# Add paths for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "angzarr"))
sys.path.insert(0, str(Path(__file__).parent.parent))

from angzarr_client.proto.angzarr import types_pb2 as types
from angzarr_client.errors import CommandRejectedError
from angzarr_client.proto.examples import table_pb2 as table
from angzarr_client import new_event_book

from .state import TableState


def handle_end_hand(
    command_book: types.CommandBook,
    command_any: Any,
    state: TableState,
    seq: int,
) -> types.EventBook:
    """Handle EndHand command."""
    if not state.exists():
        raise CommandRejectedError("Table does not exist")

    if state.status != "in_hand":
        raise CommandRejectedError("No hand in progress")

    cmd = table.EndHand()
    command_any.Unpack(cmd)

    if cmd.hand_root != state.current_hand_root:
        raise CommandRejectedError("Hand root mismatch")

    # Calculate stack changes from results
    stack_changes = {}
    for result in cmd.results:
        player_hex = result.winner_root.hex()
        if player_hex not in stack_changes:
            stack_changes[player_hex] = 0
        stack_changes[player_hex] += result.amount

    event = table.HandEnded(
        hand_root=cmd.hand_root,
        stack_changes=stack_changes,
        ended_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )
    event.results.extend(cmd.results)

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.googleapis.com/")

    return new_event_book(command_book, seq, event_any)

File: examples/rust/table/agg/src/handlers/end_hand.rs
//! EndHand command handler.

use std::collections::HashMap;

use angzarr_client::proto::examples::{EndHand, HandEnded};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::TableState;

fn guard(state: &TableState) -> CommandResult<()> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Table does not exist"));
    }
    if state.status != "in_hand" {
        return Err(CommandRejectedError::new("No hand in progress"));
    }
    Ok(())
}

fn validate(cmd: &EndHand, state: &TableState) -> CommandResult<()> {
    if hex::encode(&cmd.hand_root) != hex::encode(&state.current_hand_root) {
        return Err(CommandRejectedError::new("Hand root mismatch"));
    }
    Ok(())
}

fn compute(cmd: &EndHand) -> HandEnded {
    let mut stack_changes: HashMap<String, i64> = HashMap::new();
    for result in &cmd.results {
        let winner_hex = hex::encode(&result.winner_root);
        *stack_changes.entry(winner_hex).or_insert(0) += result.amount;
    }

    HandEnded {
        hand_root: cmd.hand_root.clone(),
        results: cmd.results.clone(),
        stack_changes,
        ended_at: Some(angzarr_client::now()),
    }
}

pub fn handle_end_hand(
    command_book: &CommandBook,
    command_any: &Any,
    state: &TableState,
    seq: u32,
) -> CommandResult<EventBook> {
    let cmd: EndHand = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    guard(state)?;
    validate(&cmd, state)?;

    let event = compute(&cmd);
    let event_any = pack_event(&event, "examples.HandEnded");

    Ok(new_event_book(command_book, seq, event_any))
}
